### DISCUSSION TITLE
Add in "Object Attributes" to control a Shader behavoir. --Feature request.

### DISCUSSION THREAD

**Discussion by Feral3d** [UPBGE/upbge#1547] [Ideas/Features for UPBGE 0.2.5+]
It would be invaluable for my game project if this feature was added to version 0.2.5 and beyond:
Currently UPBGE has the Amazing feature where you can ADD LAMP DATA in the nodes:
![image](https://user-images.githubusercontent.com/67176876/132110695-cc426a93-010d-4918-be64-7971fcd25fe3.png)

This is an amazing feature that allows me to control the shader in unique ways. Unfortunately, this also requires me to add in a deactivated light in the scene, in order to use it.

Would it be possible to also include a node to ADD OBJECT DATA? 
Object data currently exists already, however this node only pertains to the single mesh object that is using that data. 
This would allow me to to name a specific object in the scene that would control all objects with that material.

It would look something like this:
![image](https://user-images.githubusercontent.com/67176876/132110819-47d8f469-d906-43c4-b133-9606cebc8853.png)

I imagine this something that would be easy to do, as the node already exists in "Cycles". Under the texture coordinates node, where users can reference a separate object to get its coordinates:

![image](https://user-images.githubusercontent.com/67176876/132110883-445ec51e-7034-45a8-b404-6614f36e86ed.png)

The addition of this feature is also essential, because if users are using custom GLSL fragment shaders, they need to take stock of all the lamps in the scene. Allowing users to designate objects, rather than lights, to effect materials would drastically improve performance, because the custom shader would not pick up any unnecessary lamps. 

Best,
Thomas Murphy

**Comment by BluePrintRandom** [comment]
so, it's very tricky to work on new glsl nodes

however here is the glsl 

vec3 output = target_object.matrix.inverted() * fragment.xyz ; 


------------------------------------------------------------------------
and another one that would be handy too 

vec3 output = target.object.matrix.inverted() * input_vector;
#get a point in world space relative to a object

**Comment by youle31** [comment]
I think it only concerns 0.2.5 as 0.3 has already such kindof things working.

**Comment by youle31** [comment]
Converted to Discussions as it is a feature request and not a bug.

**Comment by HolandonHdunto** [comment]
But UPBGE 0.3 is not suitable for complex games, partly because you can't reduce graphics, materials are heavy ​​and there are some bugs like this https://youtu.be/_tPeXkjodDU, you can add an object with a vertex without material and the game has an absurd drop in fps, something that doesn't happen in upbge 0.2.5.
 I think the most important thing about a game engine is efficiency.

**Comment by BluePrintRandom** [comment]
This is a result of the depsgraph I think,

I have a example of a bullet pool script if you would like,
It efficiently adds bullets after the first N bullets (more players with a
higher fir rate will need a larger pool)
